### PR TITLE
Replace all CompletableFutures with PlainActionFutures

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
Replace all CompletableFutures with PlainActionFutures
Except 
- `CompletableFuture.runAsync`

### Issues Resolved
Fixes https://github.com/opensearch-project/flow-framework/issues/409

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by owaiskazi19** [comment]
@zane-neo mind taking a look.

**Review by dbwiddis — COMMENTED** [review]
LGTM!   A few suggestions:
 - We use the completed future for `WorkflowData.EMPTY` so often we should make it a constant (probalby in the `WorkflowData` class
 - You can really streamline the `ProcessNode` collection of predecessor future values.  The previous multi-step was only there to enable a varargs array for `allOf()`.
 - Replace all `future.isDone()` with a test for `future.getState()` matching `COMPLETED`.
 - Double-check the exception handling everywhere you've replaced `join()` with either `get()` or `actionGet()` as it's different.

File: src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java [hunk 1]
```diff
@@ -139,51 +141,53 @@ public TimeValue nodeTimeout() {
      * @return this node's future.
      * This is returned immediately, while process execution continues asynchronously.
      */
-    public CompletableFuture<WorkflowData> execute() {
+    public PlainActionFuture<WorkflowData> execute() {
         if (this.future.isDone()) {
             throw new IllegalStateException("Process Node [" + this.id + "] already executed.");
         }
 
         CompletableFuture.runAsync(() -> {
```
Comment (zane-neo) [inline src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java:149]: In fact I don't have preference between `CompletableFuture` and `PlainActionFuture`, the bottleneck I meant is that the runAsync or other run methods in CompletableFuture uses ForkjoinPool which could be a small number, we should provide a customer thread pool for CompletableFuture to run tasks, e.g. `CompletableFuture.runAsync(() -> {yourTask}, Executors.myExcutor())`

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java:149]: We do use our own thread pool executor for the [two times we use `runAsync()`](https://github.com/search?q=repo%3Aopensearch-project%2Fflow-framework%20runasync&type=code)
https://github.com/opensearch-project/flow-framework/blob/bf2cad068996c4aee9f38352ef3dab4a04c1823b/src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java#L188

and 

https://github.com/opensearch-project/flow-framework/blob/bf2cad068996c4aee9f38352ef3dab4a04c1823b/src/main/java/org/opensearch/flowframework/workflow/AbstractRetryableWorkflowStep.java#L131

Are there any other bottlenecks you see?

The main benefit of this replacement, as I understood it, is removing (blocking) `.join()` or `.get()` calls from `CompletableFuture` with their equivalent methods in `PlainActionFuture` which extends `ActionListener` and thus uses the `onResponse` and `onFailure` callbacks which I presume work much better in a multithreaded/multichannel environment.

File: src/test/java/org/opensearch/flowframework/workflow/CreateConnectorStepTests.java [hunk 2]
```diff
@@ -130,7 +130,7 @@ public void testCreateConnectorFailure() throws IOException {
 
         verify(machineLearningNodeClient).createConnector(any(MLCreateConnectorInput.class), any());
 
-        assertTrue(future.isCompletedExceptionally());
+        assertTrue(future.isDone());
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/test/java/org/opensearch/flowframework/workflow/CreateConnectorStepTests.java:133]: `getState()` is not a public method.

File: src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java (hunk 1 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java:147]: @zane-neo any other suggestion?

File: .github/workflows/test_security.yml [hunk 3]
```diff
@@ -16,7 +16,7 @@ jobs:
   integ-test-with-security-linux:
     strategy:
       matrix:
-        java: [11, 17, 21]
+        java: [11, 17,  21.0.1]
```
Comment (dbwiddis) [inline .github/workflows/test_security.yml:19]: Yay, security tests pass now!

File: src/test/java/org/opensearch/flowframework/workflow/CreateConnectorStepTests.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/test/java/org/opensearch/flowframework/workflow/CreateConnectorStepTests.java:133]: I guess we're stuck with `isDone()`.

**Review by dbwiddis — APPROVED** [review]
LGTM!

File: src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java [hunk 4]
```diff
@@ -139,51 +139,45 @@ public TimeValue nodeTimeout() {
...
-            List<CompletableFuture<WorkflowData>> predFutures = predecessors.stream().map(p -> p.future()).collect(Collectors.toList());
             try {
-                if (!predecessors.isEmpty()) {
-                    CompletableFuture<Void> waitForPredecessors = CompletableFuture.allOf(predFutures.toArray(new CompletableFuture<?>[0]));
-                    waitForPredecessors.join();
-                }
-
                 logger.info("Starting {}.", this.id);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/workflow/ProcessNode.java:149]: Already approved but one quick note. This log entry should be after the `actionGet()` call so it doesn't post until the previous nodes have all finished.

**Comment by dbwiddis** [comment]
Integ tests are failing in JDK 21, probably need all our CI updated to force 21.0.1

File: .github/workflows/test_security.yml (hunk 3 above)
Comment (dbwiddis) [reply .github/workflows/test_security.yml:19]: I've applied this change in #429 so you can just rebase to main and leave it unchanged

Comment (joshpalis) [reply .github/workflows/test_security.yml]: PR https://github.com/opensearch-project/flow-framework/pull/432 fixes our multi-node integration tests, perhaps lets merge that PR in first prior to this PR so we can gauge if these changes have any effect.

**Review by joshpalis — APPROVED** [review]
Changes LGTM

**Pull request merged**